src/metric.cc

```diff
@@ -0,0 +1,285 @@
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "metric.h"
+
+#ifdef TRITON_PB_STUB
+#include "pb_stub.h"
+#endif
+
+namespace triton { namespace backend { namespace python {
+
+Metric::Metric(const std::string& labels, void* metric_family_address)
+    : labels_(labels), operation_value_(0), metric_address_(nullptr),
+      metric_family_address_(metric_family_address), is_cleared_(false)
+{
+#ifdef TRITON_PB_STUB
+  SendCreateMetricRequest();
+#endif
+}
+
+Metric::~Metric()
+{
+#ifdef TRITON_PB_STUB
+  Clear();
+#endif
+}
+
+void
+Metric::SaveToSharedMemory(std::unique_ptr<SharedMemoryManager>& shm_pool)
+{
+  AllocatedSharedMemory<MetricShm> custom_metric_shm =
+      shm_pool->Construct<MetricShm>();
+  custom_metric_shm_ptr_ = custom_metric_shm.data_.get();
+
+  std::unique_ptr<PbString> labels_shm = PbString::Create(shm_pool, labels_);
+
+  custom_metric_shm_ptr_->operation_value = operation_value_;
+  custom_metric_shm_ptr_->labels_shm_handle = labels_shm->ShmHandle();
+  custom_metric_shm_ptr_->metric_family_address = metric_family_address_;
+  custom_metric_shm_ptr_->metric_address = metric_address_;
+
+  // Save the references to shared memory.
+  custom_metric_shm_ = std::move(custom_metric_shm);
+  labels_shm_ = std::move(labels_shm);
+  shm_handle_ = custom_metric_shm.handle_;
+}
+
+std::unique_ptr<Metric>
+Metric::LoadFromSharedMemory(
+    std::unique_ptr<SharedMemoryManager>& shm_pool,
+    bi::managed_external_buffer::handle_t handle)
+{
+  AllocatedSharedMemory<MetricShm> custom_metric_shm =
+      shm_pool->Load<MetricShm>(handle);
+  MetricShm* custom_metric_shm_ptr = custom_metric_shm.data_.get();
+
+  std::unique_ptr<PbString> labels_shm = PbString::LoadFromSharedMemory(
+      shm_pool, custom_metric_shm_ptr->labels_shm_handle);
```

Comment:
Had a conversation with Kris offline about the performance impact of having to Save/Load the metric object on every single operation. The `labels` in particular may be costly to serialize/deserialize, especially as number of labels increases. She will run some local tests to get a rough idea of the impact. If there is a large impact, we had some potential ideas to follow-up with like communicating only the address as an identifier, or persistent shared memory for long lived objects like metrics.

Reply:
Yeah, that can be costly. I think this is a follow up that can be addressed in the next release.

Reply:
Follow-up sounds good to me, we can get this out the door to start seeing it used in the wild.

Reply:
Created a follow-up ticket for this. Let me address that in the next release. Thanks again for all the reviews and discussion!